simpleclient_servlet/src/main/java/io/prometheus/client/filter/MetricsFilter.java

```diff
@@ -0,0 +1,98 @@
+ * performance.
+ *
+ * By default, this filter will provide metrics that distinguish 3 levels deep for the request path
+ * (including servlet context path), but can be configured with the {@code path-components} init parameter.
+ *
+ * The Histogram buckets can be configured with a {@code buckets} init parameter whose value is a comma-separated list of valid {@code double} values.
+ *
+ * The Histogram name itself can also be configured with a {@code metric-name} init parameter, whose value will be used directly.
+ */
+public class MetricsFilter implements Filter {
+    public static final String PATH_COMPONENT_PARAM = "path-components";
+    public static final String METRIC_NAME_PARAM = "metric-name";
+    public static final String BUCKET_CONFIG_PARAM = "buckets";
+    public static final String DEFAULT_FILTER_NAME = "servlet_request_latency_seconds";
+    public static final int DEFAULT_PATH_COMPONENTS = 3;
+
+    private static Histogram servletLatency = null;
+
+    // Package-level for testing purposes
+    int pathComponents = DEFAULT_PATH_COMPONENTS;
+
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+        Histogram.Builder builder = Histogram.build()
+                .help("The time taken fulfilling servlet requests")
+                .labelNames("path", "method");
+
+        if (filterConfig == null) {
+            servletLatency = builder.name(DEFAULT_FILTER_NAME).register();
+            return;
+        }
+
+        // Allow custom metric naming, falling back to default name
+        builder.name(StringUtils.defaultIfEmpty(filterConfig.getInitParameter(METRIC_NAME_PARAM), DEFAULT_FILTER_NAME));
+
+        // Allow overriding of the path "depth" to track
+        if (!StringUtils.isEmpty(filterConfig.getInitParameter(PATH_COMPONENT_PARAM))) {
+            pathComponents = Integer.valueOf(filterConfig.getInitParameter(PATH_COMPONENT_PARAM));
+        }
+
+        // Allow users to override the default bucket configuration
+        if (!StringUtils.isEmpty(filterConfig.getInitParameter(BUCKET_CONFIG_PARAM))) {
+            String[] bucketParams = filterConfig.getInitParameter(BUCKET_CONFIG_PARAM).split(",");
+            double[] buckets = new double[bucketParams.length];
+
+            for (int i = 0; i < bucketParams.length; i++) {
+                buckets[i] = Double.parseDouble(bucketParams[i]);
+            }
+
+            builder.buckets(buckets);
+        }
+
+        servletLatency = builder.register();
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain) throws IOException, ServletException {
+        if (!(servletRequest instanceof HttpServletRequest)) {
+            filterChain.doFilter(servletRequest, servletResponse);
+            return;
+        }
+
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+
+        String path = request.getRequestURI();
```

Comment:
I have a similar filter that I'm using that does the following:
```
String handlerUrl = (String) req.getAttribute(HandlerMapping.BEST_MATCHING_PATTERN_ATTRIBUTE);
if(handlerUrl == null){
  handlerUrl = req.getRequestURI();
}
```

This way the URI matches the `RequestMapping` definition that was chosen. Unfortunately it means you would have to look it up after calling `doFilter` and allowing Spring to add the attribute.

I also will allow you to avoid the path components/matching approach.